In [1]:
from Skiier_revised import *
import pandas as pd
import time
from tqdm import tqdm

### 실제 정답
- state 10,40 -> 똑같음 찾기 어려움

In [8]:
real_solution =np.array([
  [-5.410774, -5.107744],
  [-4.441077, -4.410774],
  [-3.666667, -3.441077],
  [-2.666667, -3.344108],
  [-1.666667, -1.666667],
  [-2.000000, -1.666667],
  [-1.000000, -1.666667],
  [ 0.000000,  0.000000]] )


In [28]:
Q = np.zeros([real_solution.shape[0],real_solution.shape[1]] )
argmax=np.argmax(real_solution, axis=1)
   

for i in range(len(Q)):
    if argmax[i] ==1 :
        Q[i] = [0,1]
    elif argmax[i] ==0 :
        Q[i] = [1,0]
        
print(pd.DataFrame(Q, columns=['n','s'], index =np.arange(0,80,10)).T)

    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0


### Policy iteration 1 - MC Control 

### Grid Search

In [58]:
exploration_decay =np.arange(0.989, 0.9999, 0.0005)
learning_rate = np.arange(0.005, 0.02, 0.001)

In [59]:
learning_rate

array([0.005, 0.006, 0.007, 0.008, 0.009, 0.01 , 0.011, 0.012, 0.013,
       0.014, 0.015, 0.016, 0.017, 0.018, 0.019])

In [51]:
exploration_decay

array([0.989 , 0.9895, 0.99  , 0.9905, 0.991 , 0.9915, 0.992 , 0.9925,
       0.993 , 0.9935, 0.994 , 0.9945, 0.995 , 0.9955, 0.996 , 0.9965,
       0.997 , 0.9975, 0.998 , 0.9985, 0.999 , 0.9995])

In [53]:

for decay in tqdm(exploration_decay)  :
    for lr in learning_rate :
        beg_time =time.time()
        num_ep = 10**5
        q_s_a = q_s_a_init
        pi = pi_50
        exploration_rate = 1

        for epi_i in range(1,num_ep) :
            sample_path_i = simul_path(pi, P_normal, P_speed, R_s_a)
            q_s_a = pol_eval_MC(sample_path_i, q_s_a, alpha = lr)
            pi = pol_imp(pi, q_s_a, exploration_rate)
            exploration_rate *= decay  # exponential decay
    
        end_time =time.time()
        result_q = pd.DataFrame(q_s_a, columns =['n','s'], index= states)
        result_pi = pd.DataFrame(pi, columns =['n','s'], index= states)
        print("Time difference of {} sec".format(end_time- beg_time))
        print("-------------")
        print(result_pi.T)
        print("-------------")
        print(result_q.T)
        print("-------------");
        print("MSE :",((q_s_a-real_solution)**2).mean(), " decaying : ",decay, "learning_rate : ", lr)
        print("\n")
       


  0%|                                                                                           | 0/22 [00:00<?, ?it/s]

Time difference of 22.198956966400146 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.477877 -4.534605 -3.676550 -2.629453 -1.825229 -1.763590 -1.000000  0.0
s -5.133606 -4.341430 -3.430027 -2.755503 -1.680985 -1.817711 -1.000656  0.0
-------------
MSE : 0.05715065422109222  decaying :  0.989 learning_rate :  0.005


Time difference of 22.714230060577393 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.468926 -4.644617 -3.716892 -2.737651 -1.807508 -1.781806 -1.00000  0.0
s -5.126061 -4.475937 -3.472022 -2.784129 -1.770660 -1.806346 -1.00238  0.0
-------------
MSE : 0.05690838236111528  decaying :  0.989 learni

  5%|███▋                                                                            | 1/22 [06:02<2:06:59, 362.84s/it]

Time difference of 26.58487343788147 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  1.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.878469 -4.866431 -3.646986 -2.644174 -2.115951 -1.973484 -1.000000  0.0
s -5.250207 -4.845018 -4.061461 -3.082099 -1.644174 -2.115951 -1.040404  0.0
-------------
MSE : 0.11625458948372688  decaying :  0.989 learning_rate :  0.019


Time difference of 23.417613744735718 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.387230 -4.544409 -3.693680 -2.657734 -1.723328 -1.779452 -1.000000  0.0
s -5.120146 -4.443208 -3.494227 -2.766280 -1.788728 -1.723328 -1.002096  0.0
-------------
MSE : 0.05384834393800945  decaying :  0.98949999

  9%|███████▎                                                                        | 2/22 [12:17<2:03:17, 369.90s/it]

Time difference of 30.833238124847412 sec
-------------
    0    10   20   30   40   50   60   70
n  1.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.631679 -4.631679 -3.631679 -2.631679 -1.640871 -1.984181 -1.000000  0.0
s -6.018585 -4.956052 -4.013541 -3.009651 -2.008482 -1.640871 -1.009156  0.0
-------------
MSE : 0.1378032333596472  decaying :  0.9894999999999999 learning_rate :  0.019


Time difference of 27.162534475326538 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.403275 -4.528398 -3.691935 -2.639696 -1.684224 -1.775230 -1.000000  0.0
s -5.111119 -4.404864 -3.491414 -2.765786 -1.797664 -1.684192 -1.005565  0.0
-------------
MSE : 0.05321477240151025  decaying 

 14%|██████████▉                                                                     | 3/22 [18:38<1:58:44, 374.99s/it]

Time difference of 28.246663570404053 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -6.051488 -4.939963 -4.000000 -3.000000 -2.000000 -2.000000 -1.000000  0.0
s -5.600849 -5.028172 -4.014913 -3.110838 -2.069244 -2.001842 -1.008669  0.0
-------------
MSE : 0.16925886177359456  decaying :  0.9899999999999999 learning_rate :  0.019


Time difference of 22.828237056732178 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.459219 -4.576161 -3.681240 -2.693245 -1.691315 -1.804958 -1.000000  0.0
s -5.200931 -4.427401 -3.528366 -2.797347 -1.853159 -1.691315 -1.000822  0.0
-------------
MSE : 0.053401615771450306  decayin

 18%|██████████████▌                                                                 | 4/22 [24:53<1:52:30, 375.04s/it]

Time difference of 27.55129051208496 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -6.002125 -4.955471 -4.000000 -3.000000 -2.000000 -2.000000 -1.000000  0.0
s -5.600944 -4.984465 -4.041149 -3.022469 -2.050728 -2.035945 -1.004858  0.0
-------------
MSE : 0.16908729030759656  decaying :  0.9904999999999998 learning_rate :  0.019


Time difference of 23.032379388809204 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.444832 -4.587649 -3.711092 -2.676629 -1.840039 -1.799611 -1.000000  0.0
s -5.108925 -4.458905 -3.453677 -2.807599 -1.642489 -1.800566 -1.001455  0.0
-------------
MSE : 0.052891832733027264  decaying

 23%|██████████████████▏                                                             | 5/22 [30:49<1:44:18, 368.13s/it]

Time difference of 26.451237678527832 sec
-------------
    0    10   20   30   40   50   60   70
n  1.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.675983 -4.675983 -3.675983 -2.675983 -1.655217 -1.971432 -1.000000  0.0
s -6.024374 -4.902844 -4.082954 -3.001968 -1.984178 -1.655217 -1.053254  0.0
-------------
MSE : 0.13845414188768013  decaying :  0.9909999999999998 learning_rate :  0.019


Time difference of 22.199644804000854 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20       30        40        50        60   70
n -5.390704 -4.436382 -3.662641 -2.73980 -1.665472 -1.787693 -1.000000  0.0
s -5.088708 -4.592600 -3.483825 -2.76292 -1.830682 -1.665472 -1.002562  0.0
-------------
MSE : 0.05573954403381528  decaying : 

 27%|█████████████████████▊                                                          | 6/22 [36:39<1:36:30, 361.93s/it]

Time difference of 24.524421215057373 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  0.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.849239 -4.528485 -3.746236 -2.744759 -2.277114 -1.972909 -1.000000  0.0
s -5.353140 -4.849237 -3.965618 -3.039105 -1.744759 -2.277114 -1.091745  0.0
-------------
MSE : 0.11972612961864255  decaying :  0.9914999999999997 learning_rate :  0.019


Time difference of 21.835618495941162 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30       40        50        60   70
n -5.471691 -4.540446 -3.655006 -2.686137 -1.79519 -1.755591 -1.000000  0.0
s -5.183338 -4.403316 -3.508053 -2.784702 -1.72236 -1.769646 -1.000358  0.0
-------------
MSE : 0.05445115518503257  decaying : 

 32%|█████████████████████████▍                                                      | 7/22 [42:34<1:29:55, 359.73s/it]

Time difference of 23.497312307357788 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.848653 -4.889537 -3.970021 -2.679888 -2.046627 -1.977674 -1.000000  0.0
s -5.414706 -4.522236 -3.748604 -2.971289 -1.710857 -2.046627 -1.003469  0.0
-------------
MSE : 0.09726810529814797  decaying :  0.9919999999999997 learning_rate :  0.019


Time difference of 21.97026038169861 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.519878 -4.564098 -3.713842 -2.665755 -1.666287 -1.758847 -1.000000  0.0
s -5.114829 -4.442779 -3.471160 -2.751769 -1.812190 -1.666287 -1.005004  0.0
-------------
MSE : 0.05620243834573383  decaying 

 36%|█████████████████████████████                                                   | 8/22 [48:37<1:24:10, 360.78s/it]

Time difference of 24.911351919174194 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.980885 -4.853811 -4.000000 -3.000000 -2.000000 -2.000000 -1.000000  0.0
s -5.594670 -4.971434 -4.018593 -3.051685 -2.049858 -2.011702 -1.055157  0.0
-------------
MSE : 0.15243806169444998  decaying :  0.9924999999999996 learning_rate :  0.019


Time difference of 22.141520977020264 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.426930 -4.583158 -3.632133 -2.639845 -1.814796 -1.787650 -1.00000  0.0
s -5.035238 -4.346908 -3.409968 -2.810337 -1.662102 -1.805276 -1.00419  0.0
-------------
MSE : 0.05266999290337429  decaying : 

 41%|████████████████████████████████▋                                               | 9/22 [54:23<1:17:07, 356.00s/it]

Time difference of 26.047930479049683 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.861534 -4.699171 -3.730912 -2.730912 -1.709284 -1.980031 -1.000000  0.0
s -5.350746 -4.876525 -3.948497 -3.031457 -1.995873 -1.709284 -1.050154  0.0
-------------
MSE : 0.08760915093028052  decaying :  0.9929999999999995 learning_rate :  0.019


Time difference of 21.883489847183228 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.445087 -4.518910 -3.654216 -2.692293 -1.670725 -1.787293 -0.992862  0.0
s -5.182611 -4.356942 -3.496289 -2.773637 -1.846864 -1.670725 -0.995072  0.0
-------------
MSE : 0.05461716075285233  decaying

 45%|███████████████████████████████████                                          | 10/22 [1:00:22<1:11:24, 357.05s/it]

Time difference of 26.367456674575806 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  0.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.898136 -4.794761 -3.580182 -2.575845 -1.978917 -1.946665 -1.00000  0.0
s -5.185321 -4.898136 -4.042383 -2.981409 -1.575845 -1.978508 -1.20436  0.0
-------------
MSE : 0.09591047862883938  decaying :  0.9934999999999995 learning_rate :  0.019


Time difference of 21.9971923828125 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.364339 -4.542187 -3.702606 -2.641005 -1.622450 -1.753891 -1.000000  0.0
s -5.085841 -4.510682 -3.411717 -2.808541 -1.823495 -1.622450 -1.001402  0.0
-------------
MSE : 0.05275859512847028  decaying :  0

 50%|██████████████████████████████████████▌                                      | 11/22 [1:06:29<1:06:00, 360.03s/it]

Time difference of 27.086864233016968 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  1.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.857975 -4.872270 -3.737482 -2.732730 -2.158627 -1.976075 -1.000000  0.0
s -5.408418 -4.855299 -4.097296 -3.024144 -1.732730 -2.158627 -1.063559  0.0
-------------
MSE : 0.12931442572044505  decaying :  0.9939999999999994 learning_rate :  0.019


Time difference of 23.829315900802612 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.439227 -4.559550 -3.680711 -2.724308 -1.805058 -1.794190 -1.000000  0.0
s -5.275639 -4.430124 -3.614972 -2.789783 -1.771780 -1.795029 -1.004641  0.0
-------------
MSE : 0.0569848292015853  decaying 

 55%|██████████████████████████████████████████                                   | 12/22 [1:12:40<1:00:35, 363.51s/it]

Time difference of 25.309921741485596 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  0.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.880130 -4.351355 -3.998709 -3.172538 -1.580017 -1.977397 -1.000000  0.0
s -5.009063 -4.880130 -3.275225 -3.144342 -2.172538 -1.580017 -1.264229  0.0
-------------
MSE : 0.0828335710279893  decaying :  0.9944999999999994 learning_rate :  0.019


Time difference of 22.621488332748413 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.496808 -4.498654 -3.701595 -2.649927 -1.815584 -1.787369 -1.000000  0.0
s -5.253786 -4.626212 -3.531658 -2.759691 -1.658813 -1.798413 -1.013807  0.0
-------------
MSE : 0.0587962713978096  decaying :

 59%|██████████████████████████████████████████████▋                                | 13/22 [1:18:50<54:48, 365.34s/it]

Time difference of 26.35462999343872 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40       50        60   70
n -5.990383 -5.114623 -4.000000 -3.000000 -2.000000 -2.00000 -1.000000  0.0
s -5.679532 -4.996301 -4.296331 -3.166042 -2.042092 -2.04104 -1.208586  0.0
-------------
MSE : 0.190426899768724  decaying :  0.9949999999999993 learning_rate :  0.019


Time difference of 23.656904935836792 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.439007 -4.588152 -3.668098 -2.670132 -1.821285 -1.773351 -1.000000  0.0
s -5.131400 -4.487280 -3.420179 -2.800438 -1.645935 -1.801212 -1.004433  0.0
-------------
MSE : 0.05357693723155428  decaying :  0.

 64%|██████████████████████████████████████████████████▎                            | 14/22 [1:25:09<49:14, 369.33s/it]

Time difference of 26.103164196014404 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.988178 -5.110872 -4.000000 -3.000000 -2.000000 -2.000000 -1.00000  0.0
s -5.619619 -4.960123 -4.316656 -3.097073 -2.094052 -2.057105 -1.17214  0.0
-------------
MSE : 0.19290458826742768  decaying :  0.9954999999999993 learning_rate :  0.019


Time difference of 24.459070920944214 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.409308 -4.573966 -3.655833 -2.710377 -1.814337 -1.781381 -1.000000  0.0
s -5.077145 -4.454895 -3.433908 -2.789113 -1.694782 -1.792235 -1.002753  0.0
-------------
MSE : 0.05359899383561917  decaying : 

 68%|█████████████████████████████████████████████████████▊                         | 15/22 [1:31:31<43:33, 373.36s/it]

Time difference of 28.422702312469482 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -6.049282 -4.888175 -4.000000 -3.000000 -2.000000 -2.000000 -1.000000  0.0
s -5.596060 -4.982640 -4.052258 -3.034727 -2.103423 -2.054009 -1.313386  0.0
-------------
MSE : 0.1525787237269941  decaying :  0.9959999999999992 learning_rate :  0.019


Time difference of 24.222321271896362 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.430842 -4.580190 -3.655231 -2.673619 -1.629383 -1.801157 -1.000000  0.0
s -5.180925 -4.462975 -3.505492 -2.761893 -1.817440 -1.629383 -1.001632  0.0
-------------
MSE : 0.054903763571727285  decaying

 73%|█████████████████████████████████████████████████████████▍                     | 16/22 [1:37:37<37:06, 371.01s/it]

Time difference of 30.03744864463806 sec
-------------
    0    10   20   30   40   50   60   70
n  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0
s  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -6.000000 -5.000000 -4.000000 -3.000000 -2.000000 -2.000000 -1.000000  0.0
s -6.063401 -5.013652 -4.015375 -3.071972 -2.110439 -2.005492 -1.322891  0.0
-------------
MSE : 0.19396575528700175  decaying :  0.9964999999999992 learning_rate :  0.019


Time difference of 23.364456176757812 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.376565 -4.639939 -3.649262 -2.719613 -1.791306 -1.751792 -1.000000  0.0
s -5.087300 -4.477277 -3.438676 -2.779079 -1.705371 -1.768138 -1.002249  0.0
-------------
MSE : 0.05614460895262219  decaying 

 77%|█████████████████████████████████████████████████████████████                  | 17/22 [1:43:51<31:00, 372.03s/it]

Time difference of 27.705081462860107 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -6.122957 -4.873576 -4.000000 -3.000000 -2.000000 -2.000000 -1.000000  0.0
s -5.702879 -5.123395 -4.080425 -3.097652 -2.068095 -2.012894 -1.392137  0.0
-------------
MSE : 0.16971859795299016  decaying :  0.9969999999999991 learning_rate :  0.019


Time difference of 23.00182867050171 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.394373 -4.421772 -3.639835 -2.686166 -1.801108 -1.761674 -1.00000  0.0
s -4.966354 -4.606953 -3.323676 -2.777255 -1.627931 -1.794006 -1.01485  0.0
-------------
MSE : 0.057048696565919516  decaying : 

 82%|████████████████████████████████████████████████████████████████▋              | 18/22 [1:49:57<24:40, 370.22s/it]

Time difference of 24.83259916305542 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  0.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  1.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.831099 -4.316106 -3.928300 -2.921082 -1.652583 -1.950480 -1.000000  0.0
s -4.934586 -4.815595 -3.330824 -2.907942 -1.951146 -1.652583 -1.539531  0.0
-------------
MSE : 0.051354395012882505  decaying :  0.997499999999999 learning_rate :  0.019


Time difference of 22.092892169952393 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.364445 -4.602928 -3.639330 -2.652441 -1.793636 -1.774946 -1.000000  0.0
s -5.116807 -4.453279 -3.462377 -2.754084 -1.680957 -1.786267 -1.073339  0.0
-------------
MSE : 0.05081752981026738  decaying 

 86%|████████████████████████████████████████████████████████████████████▏          | 19/22 [1:56:13<18:35, 371.82s/it]

Time difference of 27.523639678955078 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
-------------
         0         10       20        30        40        50        60   70
n -6.153047 -4.868180 -4.00000 -3.000000 -2.000000 -2.000000 -1.000000  0.0
s -5.780739 -5.153047 -4.00988 -3.287388 -2.108678 -2.089472 -1.516491  0.0
-------------
MSE : 0.17462836231397486  decaying :  0.997999999999999 learning_rate :  0.019


Time difference of 23.740485191345215 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  1.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  0.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.454615 -4.575398 -3.712581 -2.656241 -1.804367 -1.781119 -1.000000  0.0
s -5.090254 -4.403222 -3.416982 -2.757181 -1.677305 -1.788019 -1.060515  0.0
-------------
MSE : 0.05104616546761568  decaying :  

 91%|███████████████████████████████████████████████████████████████████████▊       | 20/22 [2:02:41<12:33, 376.64s/it]

Time difference of 27.202516555786133 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0
s  1.0  1.0  0.0  0.0  1.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.864813 -4.986775 -3.563334 -2.558493 -2.017100 -2.067379 -1.000000  0.0
s -5.280233 -4.825730 -3.946232 -3.101348 -1.558493 -2.017100 -1.629006  0.0
-------------
MSE : 0.08160236284732322  decaying :  0.9984999999999989 learning_rate :  0.019


Time difference of 24.916338205337524 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.448374 -4.563685 -3.666385 -2.689029 -1.656628 -1.811644 -1.00000  0.0
s -5.067489 -4.504809 -3.428894 -2.796549 -1.854313 -1.656628 -1.19499  0.0
-------------
MSE : 0.03879679344377998  decaying : 

 95%|███████████████████████████████████████████████████████████████████████████▍   | 21/22 [2:09:11<06:20, 380.61s/it]

Time difference of 28.1682710647583 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.866379 -4.741378 -3.595766 -2.590964 -2.005733 -2.058357 -1.000000  0.0
s -5.291321 -4.866379 -4.154214 -3.611502 -1.590964 -2.005733 -1.792638  0.0
-------------
MSE : 0.08654945356856139  decaying :  0.9989999999999989 learning_rate :  0.019


Time difference of 24.64059805870056 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.492428 -4.579074 -3.774616 -2.630345 -1.781261 -2.009329 -1.000000  0.0
s -5.136024 -4.425083 -3.477238 -3.072377 -1.670640 -1.781261 -1.572316  0.0
-------------
MSE : 0.009381304802374002  decaying :

100%|███████████████████████████████████████████████████████████████████████████████| 22/22 [2:15:31<00:00, 369.59s/it]

Time difference of 26.368605613708496 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.785509 -4.785442 -3.797262 -2.780322 -1.972530 -2.027226 -1.000000  0.0
s -5.423814 -4.788589 -3.929935 -3.628461 -1.780322 -1.972530 -1.730235  0.0
-------------
MSE : 0.06601706954444354  decaying :  0.9994999999999988 learning_rate :  0.019




#### 실험 결과 

정답과 같게 나온 조합 

MSE : 0.05384834393800945  decaying :  0.9894999999999999 learning_rate :  0.005

MSE : 0.053464736206858175  decaying :  0.9894999999999999 learning_rate :  0.006

MSE : 0.051484285230378006  decaying :  0.9894999999999999 learning_rate :  0.007

MSE : 0.05754899517040677  decaying :  0.9894999999999999 learning_rate :  0.008

MSE : 0.053950694149495645  decaying :  0.9894999999999999 learning_rate :  0.009000000000000001

MSE : 0.0468714912101883  decaying :  0.9894999999999999 learning_rate :  0.01

MSE : 0.05360575638062484  decaying :  0.9894999999999999 learning_rate :  0.012

MSE : 0.056302162483883055  decaying :  0.9894999999999999 learning_rate :  0.014000000000000002
MSE : 0.05321477240151025  decaying :  0.9899999999999999 learning_rate :  0.005

MSE : 0.06847038978392139  decaying :  0.9899999999999999 learning_rate :  
0.018000000000000002

MSE : 0.053401615771450306  decaying :  0.9904999999999998 learning_rate :  0.005

MSE : 0.0623687014415937  decaying :  0.9909999999999998 learning_rate :  0.013000000000000001

MSE : 0.05201147552850681  decaying :  0.9914999999999997 learning_rate :  0.01

MSE : 0.053999893105639366  decaying :  0.9919999999999997 learning_rate :  0.009000000000000001

MSE : 0.06205376994442717  decaying :  0.9919999999999997 learning_rate :  0.011

MSE : 0.049606493916748054  decaying :  0.9949999999999993 learning_rate :  0.007

MSE : 0.04740661729748284  decaying :  0.9969999999999991 learning_rate :  0.016

MSE : 0.044212770816494995  decaying :  0.997999999999999 learning_rate :  0.008

MSE : 0.03736370689542816  decaying :  0.997999999999999 learning_rate :  0.009000000000000001

MSE : 0.03606687793858906  decaying :  0.9984999999999989 learning_rate :  0.007

MSE : 0.038945670136686594  decaying :  0.9984999999999989 learning_rate :  0.009000000000000001

MSE : 0.012487838559367154  decaying :  0.9989999999999989 learning_rate :  0.009000000000000001

MSE : 0.01317764920460316  decaying :  0.9989999999999989 learning_rate :  0.011

MSE : 0.014124359952948074  decaying :  0.9994999999999988 learning_rate :  0.008

MSE : 0.00734581922269165  decaying :  0.9994999999999988 learning_rate :  0.009000000000000001
.. 다수 



### 파라미터 실험
MSE : 0.00734581922269165  

decaying :  0.9995 learning_rate :  0.009

변동성이 너무 큼

-> 위의 search 결과는 운이 좋아서 그런것으로 보임

In [42]:
beg_time =time.time()
num_ep = 10**5
q_s_a = q_s_a_init
pi = pi_50
exploration_rate = 1

for epi_i in range(1,num_ep) :
    sample_path_i = simul_path(pi, P_normal, P_speed, R_s_a)
    q_s_a = pol_eval_MC(sample_path_i, q_s_a, alpha = 0.009)
    pi = pol_imp(pi, q_s_a, exploration_rate)
    exploration_rate *= 0.9995  # exponential decay
    
end_time =time.time()
result_q = pd.DataFrame(q_s_a, columns =['n','s'], index= states)
result_pi = pd.DataFrame(pi, columns =['n','s'], index= states)
print("Time difference of {} sec".format(end_time- beg_time))
print("-------------")
print(result_pi.T)
print("-------------")
print(result_q.T)
print("-------------");
print("MSE :",((q_s_a-real_solution)**2).mean(), " decaying : ",0.9995, "learning_rate : ", 0.009)
print("\n")
       

Time difference of 23.421976804733276 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  0.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0
-------------
        0         10        20        30        40        50        60   70
n -5.47616 -4.699614 -3.790846 -2.670674 -1.887589 -2.058471 -1.000000  0.0
s -5.03484 -4.344880 -3.368675 -3.425124 -1.625501 -1.883980 -1.717377  0.0
-------------
MSE : 0.013233243267497672  decaying :  0.9995 learning_rate :  0.009




###  Policy iteration 2 - TD Control 

In [61]:

for decay in tqdm(exploration_decay)  :
    for lr in learning_rate :
        beg_time =time.time()
        num_ep = 10**5
        q_s_a = q_s_a_init
        pi = pi_50
        exploration_rate = 1

        for epi_i in range(1,num_ep) :
            s_now = 0
            while s_now != 70:
                sample_step = simul_step(pi,s_now, P_normal, P_speed, R_s_a)
                q_s_a = pol_eval_TD(sample_step, q_s_a, alpha = lr)
                pi = pol_imp(pi, q_s_a, epsilon=  exploration_rate)
                s_now = int(sample_step[3])
                exploration_rate *=decay
    
        end_time =time.time()
        result_q = pd.DataFrame(q_s_a, columns =['n','s'], index= states)
        result_pi = pd.DataFrame(pi, columns =['n','s'], index= states)
        print("Time difference of {} sec".format(end_time- beg_time))
        print("-------------")
        print(result_pi.T)
        print("-------------")
        print(result_q.T)
        print("-------------");
        print("MSE :",((q_s_a-real_solution)**2).mean(), " decaying : ",decay, "learning_rate : ", lr)
        print("\n")
       


  0%|                                                                                           | 0/22 [00:00<?, ?it/s]

Time difference of 37.25150775909424 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.284639 -4.464387 -3.623924 -2.673595 -1.653964 -1.771733 -1.000000  0.0
s -5.093668 -4.504751 -3.465008 -2.700838 -1.738410 -1.648088 -1.001889  0.0
-------------
MSE : 0.05883858689132981  decaying :  0.989 learning_rate :  0.005


Time difference of 38.5605890750885 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.291599 -4.490689 -3.611172 -2.647802 -1.630978 -1.776664 -1.000000  0.0
s -5.025738 -4.415729 -3.370768 -2.688467 -1.732553 -1.649007 -1.002441  0.0
-------------
MSE : 0.05991619986730685  decaying :  0.989 learni

  5%|███▋                                                                            | 1/22 [11:22<3:58:45, 682.17s/it]

Time difference of 46.39358687400818 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.634474 -4.689121 -3.638181 -2.635210 -1.625251 -1.928731 -0.999997  0.0
s -5.308713 -4.691197 -3.814310 -2.832787 -1.842296 -1.613517 -1.004912  0.0
-------------
MSE : 0.06947071246625905  decaying :  0.989 learning_rate :  0.019


Time difference of 39.42668128013611 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.249587 -4.494035 -3.618217 -2.655320 -1.639590 -1.787523 -1.000000  0.0
s -4.991627 -4.305466 -3.401205 -2.690604 -1.742983 -1.619628 -1.006255  0.0
-------------
MSE : 0.06090953483259216  decaying :  0.989499999

  9%|███████▎                                                                        | 2/22 [22:46<3:47:45, 683.27s/it]

Time difference of 46.082425594329834 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.603020 -4.705310 -3.663312 -2.668824 -1.665842 -1.955195 -1.00000  0.0
s -5.262006 -4.706146 -3.834239 -2.847237 -1.887577 -1.638740 -1.00707  0.0
-------------
MSE : 0.0691218689784939  decaying :  0.9894999999999999 learning_rate :  0.019


Time difference of 39.074541091918945 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.295646 -4.486209 -3.626145 -2.695691 -1.670480 -1.771112 -1.000000  0.0
s -5.185994 -4.506901 -3.579575 -2.700857 -1.726941 -1.635662 -1.000263  0.0
-------------
MSE : 0.0604490917767889  decaying :  0

 14%|██████████▉                                                                     | 3/22 [33:30<3:30:40, 665.27s/it]

Time difference of 47.725019693374634 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.634053 -4.519151 -3.675943 -2.676032 -1.684992 -1.900090 -0.999900  0.0
s -5.282468 -4.688008 -3.757695 -2.793246 -1.798971 -1.679833 -1.009495  0.0
-------------
MSE : 0.06419223627451044  decaying :  0.9899999999999999 learning_rate :  0.019


Time difference of 39.59841871261597 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.287604 -4.466087 -3.620527 -2.677986 -1.682930 -1.789969 -1.000000  0.0
s -5.151116 -4.510405 -3.401571 -2.687043 -1.738659 -1.684244 -1.002006  0.0
-------------
MSE : 0.059674979501778795  decaying

 18%|██████████████▌                                                                 | 4/22 [44:21<3:17:59, 659.95s/it]

Time difference of 45.28007912635803 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.576683 -4.642509 -3.668101 -2.672917 -1.673200 -1.918625 -1.000000  0.0
s -5.275572 -4.679085 -3.802242 -2.812353 -1.847314 -1.635386 -1.021956  0.0
-------------
MSE : 0.06483944721245045  decaying :  0.9904999999999998 learning_rate :  0.019


Time difference of 37.85527324676514 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.263848 -4.449749 -3.614303 -2.664436 -1.664407 -1.768188 -1.000000  0.0
s -5.091594 -4.512222 -3.427392 -2.695356 -1.738748 -1.662748 -1.000449  0.0
-------------
MSE : 0.059926752714229654  decaying 

 23%|██████████████████▏                                                             | 5/22 [54:39<3:02:39, 644.69s/it]

Time difference of 43.33669996261597 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.585112 -4.638738 -3.699002 -2.713998 -1.719582 -1.901493 -1.000000  0.0
s -5.390751 -4.647337 -3.756749 -2.775166 -1.810197 -1.672235 -1.007311  0.0
-------------
MSE : 0.06875213009131331  decaying :  0.9909999999999998 learning_rate :  0.019


Time difference of 38.47967576980591 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.263985 -4.460330 -3.642920 -2.684100 -1.678438 -1.766593 -0.999776  0.0
s -5.081757 -4.493364 -3.393778 -2.699877 -1.726621 -1.708331 -1.002255  0.0
-------------
MSE : 0.05930905288744379  decaying :

 27%|█████████████████████▎                                                        | 6/22 [1:04:54<2:49:10, 634.43s/it]

Time difference of 43.09184694290161 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.628437 -4.548943 -3.702171 -2.703223 -1.683259 -1.937041 -1.000000  0.0
s -5.323754 -4.675434 -3.821038 -2.838235 -1.867464 -1.652695 -1.008887  0.0
-------------
MSE : 0.06600129647816225  decaying :  0.9914999999999997 learning_rate :  0.019


Time difference of 37.461997747421265 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40       50        60   70
n -5.270919 -4.460307 -3.609838 -2.673329 -1.688350 -1.76897 -1.000000  0.0
s -5.109109 -4.493487 -3.523320 -2.680175 -1.734049 -1.64148 -1.002383  0.0
-------------
MSE : 0.0611193838016836  decaying :  0

 32%|████████████████████████▊                                                     | 7/22 [1:15:21<2:38:04, 632.28s/it]

Time difference of 43.46214318275452 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.555728 -4.650228 -3.648884 -2.637209 -1.632772 -1.928673 -1.000000  0.0
s -5.302765 -4.651249 -3.792040 -2.784979 -1.816295 -1.637310 -1.008131  0.0
-------------
MSE : 0.06629728743710839  decaying :  0.9919999999999997 learning_rate :  0.019


Time difference of 38.90577507019043 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.263158 -4.441972 -3.626548 -2.683717 -1.639166 -1.760313 -0.999981  0.0
s -5.032001 -4.497578 -3.333285 -2.695358 -1.716952 -1.624541 -1.000415  0.0
-------------
MSE : 0.06099127783664377  decaying :

 36%|████████████████████████████▎                                                 | 8/22 [1:25:41<2:26:34, 628.16s/it]

Time difference of 44.791091442108154 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.590234 -4.552534 -3.690384 -2.707715 -1.733799 -1.919778 -1.000000  0.0
s -5.304865 -4.677297 -3.793881 -2.812412 -1.856069 -1.776275 -1.000002  0.0
-------------
MSE : 0.06670065350542402  decaying :  0.9924999999999996 learning_rate :  0.019


Time difference of 38.67761826515198 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.244038 -4.454787 -3.621639 -2.660476 -1.645351 -1.803761 -1.000000  0.0
s -5.101992 -4.506342 -3.466663 -2.692465 -1.748429 -1.648529 -1.000812  0.0
-------------
MSE : 0.05961596484711399  decaying 

 41%|███████████████████████████████▉                                              | 9/22 [1:36:15<2:16:30, 630.06s/it]

Time difference of 45.461917877197266 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.602380 -4.521091 -3.641081 -2.644153 -1.680263 -1.925585 -1.000000  0.0
s -5.297054 -4.684124 -3.788749 -2.805873 -1.853921 -1.716884 -1.000806  0.0
-------------
MSE : 0.06575551930831712  decaying :  0.9929999999999995 learning_rate :  0.019


Time difference of 39.11447215080261 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.291875 -4.511322 -3.635872 -2.666415 -1.638685 -1.772509 -0.999663  0.0
s -5.070521 -4.413304 -3.368388 -2.689940 -1.731309 -1.613096 -1.000042  0.0
-------------
MSE : 0.05991270923588793  decaying 

 45%|███████████████████████████████████                                          | 10/22 [1:46:38<2:05:34, 627.91s/it]

Time difference of 44.9279580116272 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.671622 -4.698981 -3.661171 -2.636639 -1.608517 -1.901175 -1.000000  0.0
s -5.275115 -4.709706 -3.804332 -2.793446 -1.845143 -1.590639 -1.002929  0.0
-------------
MSE : 0.0737109297137187  decaying :  0.9934999999999995 learning_rate :  0.019


Time difference of 38.46724891662598 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.338350 -4.458875 -3.672788 -2.698709 -1.671765 -1.759083 -1.000000  0.0
s -5.115458 -4.495827 -3.459512 -2.706124 -1.724544 -1.655449 -1.000192  0.0
-------------
MSE : 0.057938478473034606  decaying : 

 50%|██████████████████████████████████████▌                                      | 11/22 [1:57:04<1:55:00, 627.27s/it]

Time difference of 45.44180250167847 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.609668 -4.670332 -3.574885 -2.557406 -1.538681 -1.944612 -1.000000  0.0
s -5.238603 -4.687000 -3.807196 -2.819930 -1.850515 -1.516354 -1.005713  0.0
-------------
MSE : 0.07046316982470487  decaying :  0.9939999999999994 learning_rate :  0.019


Time difference of 39.813079595565796 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10       20        30        40        50        60   70
n -5.342003 -4.422401 -3.65467 -2.675825 -1.686125 -1.764210 -0.999363  0.0
s -5.056997 -4.482311 -3.44232 -2.685681 -1.723128 -1.717731 -1.001783  0.0
-------------
MSE : 0.05939800724019355  decaying :  

 55%|██████████████████████████████████████████                                   | 12/22 [2:07:33<1:44:38, 627.81s/it]

Time difference of 46.45203423500061 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.621514 -4.669349 -3.655512 -2.655035 -1.651806 -1.947453 -1.000000  0.0
s -5.284264 -4.685730 -3.804971 -2.830779 -1.881555 -1.676163 -1.004924  0.0
-------------
MSE : 0.06791360127771233  decaying :  0.9944999999999994 learning_rate :  0.019


Time difference of 39.66479182243347 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  0.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.269946 -4.485878 -3.620021 -2.663478 -1.664189 -1.792749 -1.000000  0.0
s -5.052635 -4.468717 -3.378728 -2.695015 -1.751305 -1.647065 -1.000739  0.0
-------------
MSE : 0.05934978735497533  decaying :

 59%|█████████████████████████████████████████████▌                               | 13/22 [2:17:59<1:34:06, 627.34s/it]

Time difference of 46.599252700805664 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.646764 -4.545994 -3.684483 -2.681970 -1.678134 -1.939241 -1.000000  0.0
s -5.345468 -4.713771 -3.787733 -2.810952 -1.860535 -1.724467 -1.025357  0.0
-------------
MSE : 0.06725147644986429  decaying :  0.9949999999999993 learning_rate :  0.019


Time difference of 40.8467001914978 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.275633 -4.464054 -3.596580 -2.674023 -1.656220 -1.778165 -0.999725  0.0
s -5.134508 -4.494969 -3.497729 -2.689817 -1.733863 -1.666869 -1.000563  0.0
-------------
MSE : 0.06002489974912644  decaying :

 64%|█████████████████████████████████████████████████                            | 14/22 [2:28:22<1:23:27, 625.98s/it]

Time difference of 43.35813093185425 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.642717 -4.690897 -3.614454 -2.597046 -1.573479 -1.887177 -1.000000  0.0
s -5.237757 -4.710388 -3.754400 -2.772730 -1.804417 -1.586451 -1.005718  0.0
-------------
MSE : 0.07117331422781413  decaying :  0.9954999999999993 learning_rate :  0.019


Time difference of 37.864927530288696 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.267884 -4.476712 -3.639846 -2.661090 -1.639226 -1.779828 -1.00000  0.0
s -5.053801 -4.515391 -3.393752 -2.686901 -1.734223 -1.641983 -1.00001  0.0
-------------
MSE : 0.06058039862581402  decaying :  

 68%|████████████████████████████████████████████████████▍                        | 15/22 [2:38:38<1:12:41, 623.08s/it]

Time difference of 43.72602581977844 sec
-------------
    0    10   20   30   40   50   60   70
n  1.0  0.0  1.0  1.0  1.0  0.0  1.0  1.0
s  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.610504 -4.672236 -3.715556 -2.688668 -1.649426 -1.918779 -1.000000  0.0
s -5.622366 -4.539169 -3.778088 -2.802020 -1.826424 -1.593909 -1.008662  0.0
-------------
MSE : 0.0784773663077337  decaying :  0.9959999999999992 learning_rate :  0.019


Time difference of 37.87776160240173 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.260387 -4.451614 -3.615968 -2.666531 -1.661479 -1.762634 -1.000000  0.0
s -5.058612 -4.510585 -3.390020 -2.687871 -1.733648 -1.669284 -1.001691  0.0
-------------
MSE : 0.060873927365098196  decaying :

 73%|████████████████████████████████████████████████████████                     | 16/22 [2:48:53<1:02:03, 620.59s/it]

Time difference of 42.77605986595154 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.638635 -4.664114 -3.668022 -2.668969 -1.651927 -1.894948 -1.000000  0.0
s -5.363024 -4.723496 -3.757620 -2.773568 -1.825927 -1.612963 -1.001915  0.0
-------------
MSE : 0.07323420018044882  decaying :  0.9964999999999992 learning_rate :  0.019


Time difference of 36.86834144592285 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.274041 -4.431019 -3.645980 -2.671496 -1.645216 -1.785346 -1.000000  0.0
s -5.077440 -4.473271 -3.407157 -2.689199 -1.736934 -1.647078 -1.005221  0.0
-------------
MSE : 0.05896858772554861  decaying :

 77%|█████████████████████████████████████████████████████████████                  | 17/22 [2:59:40<52:21, 628.36s/it]

Time difference of 49.43139052391052 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.588445 -4.646812 -3.629663 -2.624801 -1.623529 -1.946981 -1.000000  0.0
s -5.325443 -4.652591 -3.804814 -2.811597 -1.890190 -1.616324 -1.028472  0.0
-------------
MSE : 0.06645115133682364  decaying :  0.9969999999999991 learning_rate :  0.019


Time difference of 41.91230297088623 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.292414 -4.475221 -3.624625 -2.676211 -1.669823 -1.769913 -1.000000  0.0
s -5.174229 -4.503378 -3.533023 -2.688182 -1.727972 -1.663437 -1.002838  0.0
-------------
MSE : 0.06038187491991326  decaying :

 82%|████████████████████████████████████████████████████████████████▋              | 18/22 [3:10:43<42:36, 639.03s/it]

Time difference of 45.92365574836731 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.616628 -4.662155 -3.653077 -2.661139 -1.666996 -1.908958 -1.000000  0.0
s -5.322172 -4.667811 -3.788197 -2.810609 -1.818278 -1.687540 -1.009425  0.0
-------------
MSE : 0.06701911584760536  decaying :  0.997499999999999 learning_rate :  0.019


Time difference of 40.24627137184143 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.282105 -4.451067 -3.622505 -2.667913 -1.691095 -1.782944 -0.999998  0.0
s -5.121926 -4.497343 -3.406175 -2.678915 -1.733143 -1.678987 -1.000268  0.0
-------------
MSE : 0.06039817990827255  decaying : 

 86%|████████████████████████████████████████████████████████████████████▏          | 19/22 [3:20:43<31:21, 627.13s/it]

Time difference of 42.24511456489563 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.605526 -4.659543 -3.684369 -2.677062 -1.666237 -1.897549 -1.000000  0.0
s -5.351295 -4.683473 -3.769892 -2.776381 -1.812825 -1.676716 -1.001806  0.0
-------------
MSE : 0.07026198412003899  decaying :  0.997999999999999 learning_rate :  0.019


Time difference of 36.25053429603577 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.337025 -4.451018 -3.666956 -2.676451 -1.660789 -1.776332 -1.000000  0.0
s -5.098030 -4.529095 -3.438357 -2.692388 -1.736967 -1.664418 -1.002004  0.0
-------------
MSE : 0.05882880599285959  decaying : 

 91%|███████████████████████████████████████████████████████████████████████▊       | 20/22 [3:30:31<20:30, 615.27s/it]

Time difference of 41.6390278339386 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.614099 -4.680898 -3.671223 -2.670008 -1.656126 -1.890015 -1.000000  0.0
s -5.342207 -4.701412 -3.778084 -2.781020 -1.804356 -1.623181 -1.011565  0.0
-------------
MSE : 0.0706993061222711  decaying :  0.9984999999999989 learning_rate :  0.019


Time difference of 36.02609348297119 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.248080 -4.452390 -3.608051 -2.671173 -1.656686 -1.777579 -1.000000  0.0
s -5.082957 -4.491461 -3.396804 -2.687700 -1.721010 -1.643645 -1.000997  0.0
-------------
MSE : 0.060386284173367534  decaying : 

 95%|███████████████████████████████████████████████████████████████████████████▍   | 21/22 [3:40:20<10:07, 607.41s/it]

Time difference of 42.28607130050659 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.621248 -4.666660 -3.667852 -2.663939 -1.662482 -1.956191 -1.000000  0.0
s -5.330409 -4.684181 -3.816021 -2.854561 -1.890352 -1.677778 -1.101427  0.0
-------------
MSE : 0.0607098598182233  decaying :  0.9989999999999989 learning_rate :  0.019


Time difference of 36.681925773620605 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.254325 -4.444317 -3.614700 -2.681708 -1.661962 -1.762528 -1.000000  0.0
s -5.066171 -4.492045 -3.365867 -2.682144 -1.728307 -1.709938 -1.004495  0.0
-------------
MSE : 0.061259885398455685  decaying 

100%|███████████████████████████████████████████████████████████████████████████████| 22/22 [3:50:12<00:00, 627.85s/it]

Time difference of 41.67900347709656 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  0.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50       60   70
n -5.582991 -4.509738 -3.765818 -2.789899 -1.684302 -1.929910 -1.00000  0.0
s -5.182890 -4.644974 -3.508499 -2.801801 -1.830642 -1.656382 -1.31495  0.0
-------------
MSE : 0.03590317820872851  decaying :  0.9994999999999988 learning_rate :  0.019




### 파라미터 실험 

MSE : 0.03590317820872851  decaying :  0.9994999999999988 learning_rate :  0.019

변동성이 너무 큼 

-> 위의 search 결과는 운이 좋아서 그런것으로 보임

In [52]:
beg_time =time.time()
num_ep = 10**5
q_s_a = q_s_a_init
pi = pi_50
exploration_rate = 1

for epi_i in range(1,num_ep) :
            s_now = 0
            while s_now != 70:
                sample_step = simul_step(pi,s_now, P_normal, P_speed, R_s_a)
                q_s_a = pol_eval_TD(sample_step, q_s_a, alpha = 0.019)
                pi = pol_imp(pi, q_s_a, epsilon=  exploration_rate)
                s_now = int(sample_step[3])
                exploration_rate *=0.9994999999999988 
    
end_time =time.time()
result_q = pd.DataFrame(q_s_a, columns =['n','s'], index= states)
result_pi = pd.DataFrame(pi, columns =['n','s'], index= states)
print("Time difference of {} sec".format(end_time- beg_time))
print("-------------")
print(result_pi.T)
print("-------------")
print(result_q.T)
print("-------------");
print("MSE :",((q_s_a-real_solution)**2).mean(), " decaying : ",0.9995, "learning_rate : ", 0.019)
print("\n")
       

Time difference of 43.046430587768555 sec
-------------
    0    10   20   30   40   50   60   70
n  0.0  1.0  1.0  1.0  1.0  0.0  1.0  1.0
s  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
-------------
         0         10        20        30        40        50        60   70
n -5.638351 -4.668637 -3.660703 -2.656543 -1.647700 -1.925164 -1.000000  0.0
s -5.307431 -4.697560 -3.796381 -2.800954 -1.852994 -1.608778 -1.310761  0.0
-------------
MSE : 0.05111189729836377  decaying :  0.9995 learning_rate :  0.019


